### PR TITLE
Update qubes integration

### PR DESCRIPTION
**Pull request by marmarek**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

This builds on top of #5317, with main changes being:
1. Avoid parsing jcat (decompress, json parse) to limit data processing before signature check in dom0. Since we have a sandbox for that (VM that did downloading), extract its parts there, transfer just signature to dom0, verify metadata integrity and only then re-construct jcat back. This avoids unneeded attack surface in dom0. See https://github.com/fwupd/fwupd/commit/45f6c9a60f5412647b1a2fd545986922dab224df for details.
2. Switch to xz-compressed metadata, since that's what fwupd uses nowadays. To make it a bit more future-proof, get metadata URL from remote's config, instead of hardcoding it.
3. Several fixes and cleanups.

### CODE REVIEW

File: contrib/qubes/src/fwupd_receive_updates.py [hunk 1]
```diff
@@ -127,6 +66,117 @@ def _jcat_verification(self, file_path, file_directory):
...
+        signature_path -- absolute path to signature file
+        file_path -- absolute path to the signed file location
+        """
+        assert file_path.startswith("/"), "bad file path {file_path!r}"
+        try:
+            self._pgp_parse(signature_path)
+        except Exception:
+            self.clean_cache()
+            raise
+        cmd_verify = ["sqv", signature_path, file_path, "--keyring", FWUPD_PKI_PGP]
```
Comment (DemiMarie) [inline contrib/qubes/src/fwupd_receive_updates.py:139]:
```suggestion
        cmd_verify = ["sqv", "--keyring", FWUPD_PKI_PGP, "--", signature_path, file_path]
```

File: contrib/qubes/src/qubes_fwupd_heads.py [hunk 2]
```diff
@@ -99,17 +113,22 @@ def _copy_heads_firmware(self, arch_path):
...
-            os.mkdir(HEADS_UPDATES_DIR)
+        create_dirs(HEADS_UPDATES_DIR)
         if os.path.exists(heads_update_path):
             print(f"Heads Update == {self.heads_update_version} " "already downloaded.")
             return EXIT_CODES["NOTHING_TO_DO"]
         else:
             os.mkdir(heads_boot_path)
-            shutil.copyfile(update_path, heads_update_path)
+            with tempfile.TemporaryDirectory() as tmpdir:
+                cmd_extract = ["gcab", "-x", f"--directory={tmpdir}", arch_path]
```
Comment (DemiMarie) [inline contrib/qubes/src/qubes_fwupd_heads.py:125]:
```suggestion
                cmd_extract = ["gcab", "-x", f"--directory={tmpdir}", "--", arch_path]
```

File: contrib/qubes/src/qubes_fwupdmgr.py [hunk 3]
```diff
@@ -277,17 +324,23 @@ def _read_dmi(self):
...
-        path -- absolute path of the updates files
+        arch_path -- absolute path of the update archive
         version -- version of the update
         downgrade -- downgrade flag
         """
         dmi_info = self._read_dmi()
-        path_metainfo = os.path.join(path, "firmware.metainfo.xml")
-        tree = ET.parse(path_metainfo)
+        with tempfile.TemporaryDirectory() as tmpdir:
+            cmd_extract = ["gcab", "-x", f"--directory={tmpdir}", arch_path]
```
Comment (DemiMarie) [inline contrib/qubes/src/qubes_fwupdmgr.py:337]:
```suggestion
            cmd_extract = ["gcab", "-x", f"--directory={tmpdir}", "--", arch_path]
```

File: contrib/qubes/src/vms/fwupd_download_updates.py [hunk 4]
```diff
@@ -66,21 +52,52 @@ def _download_metadata_jcat(self):
...
+        if info_id_line:
+            info_id = info_id_line[0].split(":", 1)[1].strip()
+        else:
+            info_id = None
+        if info_id and info_id != os.path.basename(metadata_url):
+            # fetch the file referenced in jcat, to workaround CDN being few
+            # hours out of sync
+            self.custom_url = os.path.dirname(metadata_url) + "/" + info_id
+        cmd_export = ["jcat-tool", "export",
+            f"--prefix={FWUPD_VM_METADATA_DIR}/", f"{self.metadata_file}.jcat"]
```
Comment (DemiMarie) [inline contrib/qubes/src/vms/fwupd_download_updates.py:69]:
```suggestion
        cmd_export = ["jcat-tool",
            f"--prefix={FWUPD_VM_METADATA_DIR}/", "--", "export", f"{self.metadata_file}.jcat"]
```

File: contrib/qubes/src/vms/fwupd_download_updates.py [hunk 5]
```diff
@@ -90,10 +107,9 @@ def download_updates(self, url, sha):
         url - url address of the update
         """
         self.validate_vm_dirs()
-        self.arch_name = url.replace("https://fwupd.org/downloads/", "")
-        self._decrypt_update_url(url)
+        self.arch_name = os.path.basename(url)
         update_path = os.path.join(FWUPD_VM_UPDATES_DIR, self.arch_name)
-        cmd_update = ["wget", "-O", update_path, self.dec_url]
+        cmd_update = ["wget", "-O", update_path, url]
```
Comment (DemiMarie) [inline contrib/qubes/src/vms/fwupd_download_updates.py:112]: Use `curl` instead?
```suggestion
        cmd_update = ["curl", "--proto", "=https", "--proto-redir", "=https", "--tlsv1.2", "--http1.1", "-sSfL", "-o", update_path, "--", url]
```

**Review by DemiMarie — COMMENTED** [review]
See inline comments.  Main suggestion is to recreate the armored signature (as builder-github v2 does) rather than trying to validate the untrusted armor.

File: contrib/qubes/src/fwupd_receive_updates.py [hunk 6]
```diff
@@ -127,6 +66,117 @@ def _jcat_verification(self, file_path, file_directory):
...
+            if lines[0:2] != ["-----BEGIN PGP SIGNATURE-----", ""]:
+                raise Exception("pgp: invalid header")
+            if lines[-1] != "-----END PGP SIGNATURE-----":
+                raise Exception("pgp: invalid footer")
+            checksum = lines[-2]
+            if checksum[0] != '=' or len(checksum) != 5:
+                raise Exception("pgp: invalid checksum format")
+            base64_data = ''.join(lines[2:-2])
+            data = base64.b64decode(base64_data)
+            crc = base64.b64decode(checksum[1:])
```
[the remark this answers is not in the document]
Comment (marmarek) [reply contrib/qubes/src/fwupd_receive_updates.py:110]: strict_mode - makes sense, length check for all of the others - not much, as it's already implicitly handled (both struct, and b64decode raise an exception on invalid length)

File: contrib/qubes/src/vms/fwupd_download_updates.py (hunk 5 above)
Comment (marmarek) [reply contrib/qubes/src/vms/fwupd_download_updates.py:112]: Curl might make sense, but I don't like this strict limits on what protocol to use, as it makes it less flexible if other remotes are used, or server configuration changes. For example `--http1.1` will exclude usage of `HTTP/2`

File: contrib/fwupd.spec.in [hunk 7]
```diff
@@ -218,6 +218,8 @@ Summary: fwupd wrapper for Qubes OS - dom0 scripts
 Requires:   gcab
 Requires:   fwupd >= 1.5.7
 Requires:   libjcat >= 0.1.6
+Requires:   python3-packaging
+Requires:   sequoia-sqv
```
Comment (hughsie) [inline contrib/fwupd.spec.in:222]: this needs to be surrounded by the qubes `if` section -- we can't add those for all systems.

Comment (hughsie) [reply contrib/fwupd.spec.in:222]: Ohh it is! ignore me.

**Comment by hughsie** [comment]
Can these be squashed down a bit?

**Comment by marmarek** [comment]
I definitely can squash fixes to commits in this very PR together, that will reduce number of commits by 5 or 6. Should I take more aggressive approach?

**Comment by hughsie** [comment]
Nah, 5/6 is fine.

**Comment by marmarek** [comment]
I mean `37 - (5 or 6)`, not `5 or 6` as result

**Pull request merged**